### ISSUE TITLE
Simplify and speed up write sequence

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6733]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#6733|
|Type|enhancement|
|Created|2013-09-08 21:08:17Z|

Write now writeSequence performs two loops over the iovec passed to it, this can be trivially simplified to be one loop.

Attachments:

* [write-sequence.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/664/6645bc99973b562a1bc67c85092864593d853e95/8185519b0f4e631186a097c75c5ff91ee817d937.diff) (930 bytes) - added by Alex on 2013-09-08 21:08:31Z - 
* [write-sequence.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/664/6645bc99973b562a1bc67c85092864593d853e95/c86d5950982d1423a67fa30468572b5f7728ca4a.diff) (933 bytes) - added by Alex on 2013-09-08 21:58:49Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6733 6733
type__enhancement enhancement
reporter__Alex Alex
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__ 
time__1378674497000000 1378674497000000
changetime__1378819743000000 1378819743000000
version__None None
owner__Alex Alex
cc__lvh
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_2"></a>|@lvh set owner to Alex|
|-|-|

Hi Alex,

Thank you for this contribution to such a core component of Twisted!

Some questions:

 1. This doesn't appear to have unit tests. Perhaps there should be a regression test where iovec is a generator (and therefore exhausted after one go).
 2. Does this positively affect performance? In which implementations?

I would be more than happy to contribute to (1). I'm presuming you had some reason to make this patch, so I'm eager to listen to what you have to say about (2).

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

> This doesn't appear to have unit tests. Perhaps there should be a regression test where iovec is a generator (and therefore exhausted after one go).

`iovec` must be a `list`.  Perhaps the interface docstring should be ... emphasized ... or something ... if this isn't already clear.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> Does this positively affect performance? In which implementations?

Not in IOCP reactor since IOCP reactor has its own implementation of `writeSequence` which also has two loops and isn't changed by this patch. :)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_5"></a>|@lvh commented|
|-|-|

Well, sure. Obviously I'm only asking about places where the new code is actually used...

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|Alex commented|
|-|-|

1) Is `writeSequence` really never tested!? This patch should be behavior identical
2) Consider this a purely code-clean up patch, though I do believe this should improve performance under both PyPy and CPython,

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_7"></a>|@lvh commented|
|-|-|

I'm not saying that the behavior shouldn't be identical or that writeSequence is never tested. I'm saying that the *new* behavior of writeSequence (that is, loop over iovec once) is never tested. There is currently no test that asserts that writeSequence only iterates over iovec once. There are plenty of tests that assert that writeSequence does something :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> > > Does this positively affect performance? In which implementations?

> > Not in IOCP reactor since IOCP reactor has its own implementation of writeSequence which also has two loops and isn't changed by this patch. :)

> Well, sure. Obviously I'm only asking about places where the new code is actually used...

The comment was meant to provoke thought about whether the IOCP implementation *should* also be changed. :)

> There is currently no test that asserts that writeSequence only iterates over iovec once.

Indeed.  And there doesn't need to be.  Just as there aren't any tests that we don't repeat other non-side-effect-y code in implementations.  It's fine to improve code by making it do *less* without adding more tests except in certain rare cases.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|Alex commented|
|-|-|

IOCP should probably get the same treatment, any preference between a seperate ticket and doing it here as well? (Probably there shouldn't be copy-pasta at all, but perhaps there's a good reason?)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_10"></a>|@lvh commented|
|-|-|

I don't know anything about IOCP, but a new ticket seems prudent.

I also don't understand why this behavior change is okay without a test, but I won't pick a fight over it. (I am reminded of the Axiom loop ordering test, which confuses me even more: both are unit tests measuring things we're assuming are performance regressions, except this relies on basic Python semantics instead of two different kinds of implementation detail.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

> I also don't understand why this behavior change is okay without a test, but I won't pick a fight over it.

I don't want to have a fight but I do want to try to explain why this change is okay without a unit test.

First though, let's set aside Axiom.  The unit tests you're talking about (which tried to verify that a SQL DBM picked table iteration order in a particular way) were a little misguided.  They were written by people trying to understand the performance characteristics of SQLite3 and who had a specific application with somewhat intense performance requirements in mind.  This is somewhat different from the situation in Twisted - a proper comparison of that code to this code involves going into a lot of detail, most of it historical.

So back to `writeSequence` - the change in behavior we are talking about is a change from iterating over the input twice to once.  From your comments, I think you see this as a visible change to the interface.  Python lets an iterator know when it is being iterated.  A generator uses this information to make itself only iterable once (by being its own iterator).  Other arbitrary application code could allow multiple iteration but produce different results.

I suggest this change is safe because `writeSequence` already demands that users _not_ use these features of Python.  A `list` cannot tell when it is being iterated over (or rather, it can tell but it doesn't care).  A `list` can be iterated over multiple times and always produce the same results.  Therefore, to a _correct_ user of this interface, this is not a visible change.  To a buggy user - one that, for example, produced different sequences of bytes on each iteration - this is a change: their first sequence will now get sent, whereas previously their second sequence would get sent.

Compare this to other changes we could make that would affect buggy application code.  For example, what if we decided to use `len()` on `iovec` (say, to allocate memory for the structure needed to use `writev()`).  This would break any code that passed in a generator (or anything else without a length, or with a length different from iteration length, or with a length that varies depending on whether you've iterated the object yet or not).  Do you think you'd have the same concerns about introducing (or, if we already used it, removing!) a use of `len(iovec)`?  If so, I think you're applying a compatibility standard stronger than we are capable of offering in general.  If not, do you think there's actually a difference between this example and the change in the proposed patch?

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_12"></a>|@lvh commented|
|-|-|

My apologies for any miscommunication: this isn't about the correctness of writeSequence. This is not an externally visible change. No confusion there.

"Only iterates over iovec once" is ostensibly a good property (if it isn't, we can just bin this ticket entirely). I'm not suggesting that passing things that can't be iterated over multiple times to writeSequence is sensible in real code, nor should the unit test be about asserting that you *can* do that if you want to. I'd like to assert writeSequence does its job in one iteration. An exhaustable iterator was the easiest way I can think of doing that.

I'm considering the case if someone reverts this behavior later. I think it would be nicer to have a red buildbot telling you "no, writeSequence ought to do this in one pass", than to rely on human verification that changes to writeSequence don't cause a regression here.

I'll happily admit that it's a strange unit test, since it's not really testing correctness or making assertions about writeSequence's contract. It's just about giving you a canary that checks that property actually still holds.

Basically, I'm asserting that if property X is worth having, it's worth having an automated check that property X actually holds.
